### PR TITLE
Ib/cmssw 8 1 x/gcc530: Rivet 2.5.2 and Yoda 1.6.5

### PR DESCRIPTION
**Pull request by xjanssen**
Rivet 2.5.2 and Yoda 1.6.5 update
Builded and tested for 81X environment with gcc530

### CODE REVIEW

File: rivet.spec [hunk 1]
```diff
@@ -11,44 +11,26 @@ Patch0: rivet-1.4.0
 %endif
 
 %if "%{?cms_cxxflags:set}" != "set"
-%define cms_cxxflags -O2 -std=c++11
+%define cms_cxxflags -O2 -std=c++0x
```
Comment (davidlt) [inline rivet.spec:14]: Why `c++0x`? Technically it's fine even if you set it to `c++14` as that is what we use for CMSSW.

File: rivet.spec [hunk 2]
```diff
@@ -11,44 +11,26 @@ Patch0: rivet-1.4.0
...
-  rm -f $CONFIG_GUESS_FILE
-  cp %{_tmppath}/config.guess $CONFIG_GUESS_FILE
-  chmod +x $CONFIG_GUESS_FILE
-done
-for CONFIG_SUB_FILE in $(find $RPM_BUILD_DIR -name 'config.sub')
-do
-  rm -f $CONFIG_SUB_FILE
-  cp %{_tmppath}/config.sub $CONFIG_SUB_FILE
-  chmod +x $CONFIG_SUB_FILE
-done
```
Comment (davidlt) [inline rivet.spec:36]: Keep this unless you know that all `config.{guess,sub}` scripts are up to date and can work on _aarch64_ and _ppc64le_ architectures. You can also ping upstream to get these updated to newer ones.

File: rivet.spec [hunk 3]
```diff
@@ -11,44 +11,26 @@ Patch0: rivet-1.4.0
...
-
-./configure --disable-silent-rules --prefix=%{i} --with-boost=${BOOST_ROOT} --with-hepmc=${HEPMC_ROOT} \
-            --with-fastjet=${FASTJET_ROOT} --with-gsl=$GSL_ROOT --with-yoda=${YODA_ROOT} \
-            --with-yaml-cpp=${YAML_CPP_ROOT} \
+./configure --disable-silent-rules --prefix=%i --with-hepmc=$HEPMC_ROOT \
+            --with-fastjet=$FASTJET_ROOT --with-gsl=$GSL_ROOT --with-yoda=${YODA_ROOT} \
```
Comment (davidlt) [inline rivet.spec:22]: Please, keep consistent style, i.e. `$HEPMC_ROOT` vs. `${HEPMC_ROOT}`.

File: rivet.spec [hunk 4]
```diff
@@ -11,44 +11,26 @@ Patch0: rivet-1.4.0
...
-            PYTHONPATH=${CYTHON_ROOT}/${PYTHON_LIB_SITE_PACKAGES} \
+            PYTHONPATH=${CYTHON_ROOT}/lib/python@PYTHONV@/site-packages \
             CXX="$(which %cms_cxx)" CXXFLAGS="%cms_cxxflags" CPPFLAGS="-I${BOOST_ROOT}/include"
 
 # The following hack insures that the bins with the library linked explicitly
 # rather than indirectly, as required by the gold linker
 perl -p -i -e "s|LIBS = $|LIBS = -lHepMC|g" bin/Makefile
 %build
-make %{makeprocesses} all PYTHONPATH=${CYTHON_ROOT}/${PYTHON_LIB_SITE_PACKAGES}
+make %makeprocesses all PYTHONPATH=${CYTHON_ROOT}/lib/python@PYTHONV@/site-packages
```
Comment (davidlt) [inline rivet.spec:31]: Please, revert this change.

```
## INITENV SETV PYTHON_LIB_SITE_PACKAGES lib/python%{python_major_version}/site-packages
```

File: rivet.spec [hunk 5]
```diff
@@ -11,44 +11,26 @@ Patch0: rivet-1.4.0
...
 %install
-make install PYTHONPATH=${CYTHON_ROOT}/${PYTHON_LIB_SITE_PACKAGES}
+make install PYTHONPATH=${CYTHON_ROOT}/lib/python@PYTHONV@/site-packages
```
Comment (davidlt) [inline rivet.spec:33]: Same here, revert.

File: yoda-toolfile.spec [hunk 6]
```diff
@@ -15,7 +15,9 @@ cat << \EOF_TOOLFILE >%i/etc/scram.d/yoda.xml
     <environment name="LIBDIR" default="$YODA_BASE/lib"/>
     <environment name="INCLUDE" default="$YODA_BASE/include"/>
   </client>
-  <use name="boost"/>
+  <use name="python"/>
+  <use name="cython"/>
+  <use name="root_cxxdefaults"/>
```
Comment (davidlt) [inline yoda-toolfile.spec:20]: Well, having `root_cxxdefaults` as the last is good, but you also need to specify which part of ROOT is being used. Look at yoda logs and figure out what exactly is being used from ROOT.

File: yoda.spec [hunk 7]
```diff
@@ -11,7 +11,7 @@ Requires: boost python cython
 %prep
 %setup -q -n %{n}/%{realversion}
 
-./configure --prefix=%i --with-boost=${BOOST_ROOT} CXX="$(which %cms_cxx)" CXXFLAGS="%cms_cxxflags"
+./configure --prefix=%i CXX="$(which %cms_cxx)" CXXFLAGS="%cms_cxxflags" --enable-root
```
Comment (davidlt) [inline yoda.spec:14]: Hopefully it doesn't pick ROOT from system standard location is one is available. Is there a way to pass a ROOT directory, e.g. `--with-root=<..>`?

**Review by davidlt — CHANGES_REQUESTED** [review]

File: rivet.spec (hunk 1 above)
Comment (xjanssen) [reply rivet.spec:14]: ouch, some of the changes here and below happened in a way I don't understand as I did not did them explicitly ... gosh

File: yoda.spec (hunk 7 above)
Comment (xjanssen) [reply yoda.spec:14]: Nope, the configure script is calling 'root-config' to get root variables

Comment (xjanssen) [reply yoda.spec:14]: I checked the log and it takes the correct root version from where I have done the build:
/afs/cern.ch/work/x/xjanssen/cms/Rivet/81X_gcc530/a/slc6_amd64_gcc530/lcg/root/6.06.08-dpekbn

File: yoda-toolfile.spec (hunk 6 above)
Comment (xjanssen) [reply yoda-toolfile.spec:20]: Hmm, these are the library of root being passed at compilation time:
 -lCore -lRIO -lNet -lHist -lGraf -lGraf3d -lGpad -lTree -lRint -lPostscript -lMatrix -lPhysics -lMathCore -lThread -lMultiProc -pthread -lm -ldl -rdynamic

This said it create a non dynamic executable 'yoda2root' which does not seems to have any dependencies at the end:

-bash-4.1$ ldd /afs/cern.ch/work/x/xjanssen/cms/Rivet/81X_gcc530/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/bin/yoda2root
    not a dynamic executable

Comment (davidlt) [reply yoda-toolfile.spec:20]: Could you share the build log?
Could you check all executables to make sure that none are dynamically linked to ROOT? I think, we don't ship static ROOT libs, but you should check.
Maybe it just records ROOT location for future use?

Please, investigate.

Comment (xjanssen) [reply yoda-toolfile.spec:20]: ~xjanssen/public/yoda.log

Comment (davidlt) [reply yoda-toolfile.spec:20]: Looks like the following library links to ROOT (`build/lib.linux-x86_64-2.7/yoda/rootcompat.so`)

Comment (xjanssen) [reply yoda-toolfile.spec:20]: -bash-4.1$ ldd opt/cmssw/slc6_amd64_gcc530/external/yoda/1.6.5/lib/python2.7/site-packages/yoda/rootcompat.so
    linux-vdso.so.1 =>  (0x00007ffc37af8000)
    libstdc++.so.6 => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/external/gcc/5.3.0/lib64/libstdc++.so.6 (0x00007efdc1c05000)
    libYODA.so => /afs/cern.ch/work/x/xjanssen/cms/Rivet/81X_gcc530/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libYODA.so (0x00007efdc1b78000)
    libpython2.7.so.1.0 => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libpython2.7.so.1.0 (0x00007efdc1972000)
    libCore.so => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libCore.so (0x00007efdc1591000)
    libRIO.so => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libRIO.so (0x00007efdc12fc000)
    libNet.so => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libNet.so (0x00007efdc1234000)
    libHist.so => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libHist.so (0x00007efdc0eae000)
    libGraf.so => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libGraf.so (0x00007efdc0d6d000)
    libGraf3d.so => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libGraf3d.so (0x00007efdc0cd2000)
    libGpad.so => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libGpad.so (0x00007efdc0c28000)
    libTree.so => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libTree.so (0x00007efdc0ae4000)
    libRint.so => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libRint.so (0x00007efdc0abe000)
    libPostscript.so => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libPostscript.so (0x00007efdc0a5e000)
    libMatrix.so => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libMatrix.so (0x00007efdc090c000)
    libPhysics.so => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libPhysics.so (0x00007efdc08c6000)
    libMathCore.so => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libMathCore.so (0x00007efdc0741000)
    libThread.so => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libThread.so (0x00007efdc0716000)
    libMultiProc.so => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libMultiProc.so (0x00007efdc0706000)
    libm.so.6 => /lib64/libm.so.6 (0x00007efdc0452000)
    libdl.so.2 => /lib64/libdl.so.2 (0x00007efdc024d000)
    libPyROOT.so => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libPyROOT.so (0x00007efdc0198000)
    libgcc_s.so.1 => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/external/gcc/5.3.0/lib64/libgcc_s.so.1 (0x00007efdc0181000)
    libpthread.so.0 => /lib64/libpthread.so.0 (0x00007efdbff63000)
    libc.so.6 => /lib64/libc.so.6 (0x00007efdbfbcf000)
    librt.so.1 => /lib64/librt.so.1 (0x00007efdbf9c7000)
    /lib64/ld-linux-x86-64.so.2 (0x00007efdc1da2000)
    libutil.so.1 => /lib64/libutil.so.1 (0x00007efdbf7c3000)
    libpcre.so.1 => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libpcre.so.1 (0x00007efdbf780000)
    liblzma.so.5 => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/liblzma.so.5 (0x00007efdbf750000)
    libz.so.1 => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libz.so.1 (0x00007efdbf736000)
    libssl.so.10 => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libssl.so.10 (0x00007efdbf6bd000)
    libcrypto.so.10 => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libcrypto.so.10 (0x00007efdbf4db000)
    libcrypt.so.1 => /lib64/libcrypt.so.1 (0x00007efdbf2a4000)
    libfreetype.so.6 => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libfreetype.so.6 (0x00007efdbf20c000)
    libTreePlayer.so => /cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_0_pre12/external/slc6_amd64_gcc530/lib/libTreePlayer.so (0x00007efdbf0dc000)
    libgssapi_krb5.so.2 => /lib64/libgssapi_krb5.so.2 (0x00007efdbee98000)
    libkrb5.so.3 => /lib64/libkrb5.so.3 (0x00007efdbebb0000)
    libcom_err.so.2 => /lib64/libcom_err.so.2 (0x00007efdbe9ac000)
    libk5crypto.so.3 => /lib64/libk5crypto.so.3 (0x00007efdbe780000)
    libresolv.so.2 => /lib64/libresolv.so.2 (0x00007efdbe565000)
    libfreebl3.so => /lib64/libfreebl3.so (0x00007efdbe362000)
    libkrb5support.so.0 => /lib64/libkrb5support.so.0 (0x00007efdbe156000)
    libkeyutils.so.1 => /lib64/libkeyutils.so.1 (0x00007efdbdf53000)
    libselinux.so.1 => /lib64/libselinux.so.1 (0x00007efdbdd33000)

Comment (xjanssen) [reply yoda-toolfile.spec:20]: OK, I have now a local version where I reverted the mis-happening changes above and I think the root dependency is behaving as it should (the ldd was showing that it finds root librairies from the CMSSW I was using to test the builded package). So I guess I could proceed with remaking the PR or do we have something else missing ?

**Pull request closed without merging**